src/site/content/en/blog/more-capable-form-controls/index.md

```diff
@@ -0,0 +1,306 @@
+
+#### `void formStateRestoreCallback(state, mode)`
+
+Called in one of two circumstances:
+
+*   When the browser restores the state of the element (for example, after a navigation, or when the browser restarts).  The <code>mode</code> argument is <code>"restore"</code> in this case.
+*   When the browser's input-assist features such as form autofilling sets a value. The <code>mode</code> argument is <code>"autocomplete"</code> in this case.
+
+The type of the first argument depends on how the `setFormValue()` method was called. For more details, see [Restoring form state](#restoring-form-state).
+
+### Restoring form state {: #restoring-form-state }
+
+Under some circumstances—like when navigating back to a page, or restarting the browser, the brower may try to restore the state of a form to the state the user left it in.
+
+For a form-associated custom element, the restored state comes from the value(s) you pass to the `setFormValue()` method. You can call the method with a single value parameter, as shown in the [earlier examples](#set-a-value), or with two parameters:
+
+```js
+this._internals.setFormValue(value, state);
+```
+
+The `value` represents the submittable value of the control. The optional `state` parameter is an _internal_ representation of the state of the control, which can include data that doesn't get sent to the server. The `state` parameter takes the same types as the `value` parameter—it can be a string, `File`, or `FormData` object.
+
+The `state` parameter is useful when you can't restore a control's state based on the value alone. For example, suppose you create a color picker with multiple modes: a  palette or an RGB color wheel. The submittable _value_ would  be the selected color in a canonical form, like `"#7fff00"`. But to restore the control to a specific state, you'd also need to know which mode it was in, so the _state_ might look like `"palette/#7fff00"`.  
+
+```js
+this._internals.setFormValue(this._value, 
+    this._mode + '/' + this._value);
+```
+
+Your code would need to restore its state based on the stored state value.
+
+```js
+formStateRestoreCallback(state, mode) {
+  if (mode == 'restore') {
+    // expects a state parameter in the form 'controlMode/value'
+    [controlMode, value] = state.split('/');
+    this._mode = controlMode;
+    this._value = value;
+  }
+  // Chrome currently doesn't handle autofill for form-associated 
+  // custom elements. In the autofill case, you might need to handle
+  // a raw value.
+}
+```
+
+In the case of a simpler control (for example a number input), the value is probably sufficient to restore the control to its previous state. If you omit `state` when calling `setFormValue`, then the value is passed to `formStateRestoreCallback`. 
```

Comment:
Please add parens to the function names.

Reply:
Thanks. I think I've got all of these now.